### PR TITLE
CESM support

### PR DESCRIPTION
**Pull request by mnlevy1981**
This branch cleans up the offline driver to make it easy to set up FEISTY runs from netCDF forcing / IC files. Improvements include

1. Improved interpolation, so we no longer need daily forcing fields
2. A notebook that generates forcing files from post-processed CESM output -- a future PR to add the post-processing to this repo would be great
3. Updated default settings to match the FOSI run, with a clear example of how to override the defaults for the previous two test cases
4. performance improvements to get runtime in line with the matlab performance -- relying on `numpy` instead of `xarray` for more computations, using the `assume_sorted` flag in the `scipy` interpolation routine, and refactoring how the offline driver stores the model state throughout the run

### CODE REVIEW

**Comment by mnlevy1981** [comment]
This PR will remain a draft until I update the documentation

**Comment by mnlevy1981** [comment]
The only current test failure is

```python

Traceback (most recent call last):
  File "/home/runner/.cache/pre-commit/repopwsd7jk3/py_env-python3.10/bin/black", line 8, in <module>
    sys.exit(patched_main())
  File "/home/runner/.cache/pre-commit/repopwsd7jk3/py_env-python3.10/lib/python3.10/site-packages/black/__init__.py", line 1282, in patched_main
    patch_click()
  File "/home/runner/.cache/pre-commit/repopwsd7jk3/py_env-python3.10/lib/python3.10/site-packages/black/__init__.py", line 1268, in patch_click
    from click import _unicodefun  # type: ignore
ImportError: cannot import name '_unicodefun' from 'click' (/home/runner/.cache/pre-commit/repopwsd7jk3/py_env-python3.10/lib/python3.10/site-packages/click/__init__.py)
```

It looks like this is due to an incompatibility between `black` and `click`, and has [been fixed](https://github.com/psf/black/issues/2964) but it's not clear to me how the environment used by `pre-commit` in [`linting.yaml`](https://github.com/marbl-ecosys/feisty/blob/main/.github/workflows/linting.yaml) is created so I don't know where to either pin `click` to an old version or update to the newest `black`

**Comment by mnlevy1981** [comment]
Assuming the testing passes (except the failure due to `black` / `click`), this is ready to be merged and I believe the documentation will be updated once this is available on `main`.

**Comment by mnlevy1981** [comment]
> It looks like this is due to an incompatibility between `black` and `click`, and has [been fixed](https://github.com/psf/black/issues/2964) but it's not clear to me how the environment used by `pre-commit` in [`linting.yaml`](https://github.com/marbl-ecosys/feisty/blob/main/.github/workflows/linting.yaml)

I ran `pre-commit autoupdate` and that brought in newer versions of these packages, so we're passing CI again

File: feisty/core/ecosystem.py [hunk 1]
```diff
@@ -274,8 +276,9 @@ def is_small(self):
 class zooplankton_type(object):
     """Data structure containing zooplankton parameters."""
 
-    def __init__(self, name, **kwargs):
+    def __init__(self, glob_id, name, **kwargs):
```
Comment (matt-long) [inline feisty/core/ecosystem.py:279]: can we change `glob_id` to `global_id` or `group_id`?

File: feisty/core/interface.py [hunk 2]
```diff
@@ -53,6 +55,7 @@ def __init__(
         settings_dict={},
         fish_ic_data=None,
         benthic_prey_ic_data=None,
+        biomass_init='constant',
```
Comment (matt-long) [inline feisty/core/interface.py:58]: ensure that this plays friendly with `fish_ic_data`

File: feisty/core/interface.py [hunk 3]
```diff
@@ -53,6 +55,7 @@ def __init__(
         settings_dict={},
         fish_ic_data=None,
```
Comment (matt-long) [inline feisty/core/interface.py:56]:
```suggestion
        fish_ic_data=1e-5,
        benthic_ic_data=2e-3,
```

File: feisty/core/interface.py (hunk 2 above)
Comment (matt-long) [reply feisty/core/interface.py:58]: another option:
- `biomass_init` --> `biomass_init_file`, default `None`

```python
if biomass_init_file is not None:
   ds_ic = xr.open...
   assert varnames for all taxa in ds_ic
   # use it a
else: 
   # use fish_ic_data
   # use benthic_ic_data
   ....
```

File: feisty/core/interface.py [hunk 4]
```diff
@@ -342,8 +359,10 @@ def _compute_predation(self):
             self.biomass,
             self.food_web,
         )
-        self.tendency_data.predation_rate[:, :] = (
-            self.tendency_data.predation_flux.data / self.biomass.isel(group=self.ndx_fish).data
+        self.tendency_data.predation_rate[
+            :, :
+        ] = self.tendency_data.predation_flux.data / np.maximum(
+            self.biomass.data[self.ndx_fish, :], 1e-300
```
Comment (matt-long) [inline feisty/core/interface.py:365]: should `1e-300` be replaced by `eps`?

Perhaps we should also introduce a parameter for minimum biomass?

File: feisty/core/process.py [hunk 5]
```diff
@@ -229,24 +227,28 @@ def compute_rescale_zoo_consumption(
     for zoo_i in food_web.zoo_names:
 
         link_ndx = food_web.prey_link_ndx[zoo_i]
-        biomass_zoo_pred = food_web._get_biomass_zoo_pred(biomass, zoo_i)
+        biomass_zoo_pred = np.maximum(food_web._get_biomass_zoo_pred(biomass, zoo_i).data, 1e-300)
```
Comment (matt-long) [inline feisty/core/process.py:230]: use `constants.eps`

File: feisty/core/process.py [hunk 6]
```diff
@@ -229,24 +227,28 @@ def compute_rescale_zoo_consumption(
...
 
         bio_con_zoo = biomass_zoo_pred * food_web.get_consumption(consumption_rate_link, prey=zoo_i)
-        bio_con_zoo_sum = bio_con_zoo.sum('group')
+        bio_con_zoo_sum = np.maximum(np.sum(bio_con_zoo, axis=0), 1e-300)
```
Comment (matt-long) [inline feisty/core/process.py:233]: use `constants.eps`

File: feisty/offline_driver.py [hunk 7]
```diff
@@ -110,6 +109,10 @@ def __init__(
         settings_in={},
         fish_ic_data=None,
         benthic_prey_ic_data=None,
+        biomass_init='constant',
+        allow_negative_forcing=False,
+        diagnostic_names=[],
```
Comment (matt-long) [inline feisty/offline_driver.py:114]: do we want to provide a utility to query available diags? Or have a minimal set of defaults?

File: feisty/offline_driver.py [hunk 8]
```diff
@@ -150,72 +154,101 @@ def __init__(
...
-            'growth_rate',
-            'reproduction_rate',
-            'recruitment_flux',
-            'fish_catch_rate',
-            'encounter_rate_link',
-            'encounter_rate_total',
-            'consumption_rate_max_pred',
-            'consumption_rate_link',
-        ]
+        self._diagnostic_names = diagnostic_names
```
Comment (matt-long) [inline feisty/offline_driver.py:160]: change `_diagnostic_names` to `_add_diagnostic_names`?

File: feisty/offline_driver.py [hunk 9]
```diff
@@ -150,72 +154,101 @@ def __init__(
...
-        ds_diag = zeros * self.obj.tendency_data[self._diagnostic_names]
-        ds_prog = zeros * self.obj.get_prognostic().to_dataset()
-        self._ds = xr.merge((ds_prog, ds_diag))
-        self.ds['time'] = self.time
+        """Create self._ds, a list of datasets containing no more than
+        self._max_output_time_dim time levels
+        """
+        self._ds = []
+        start_dates = [self.start_date]
+        days_per_year = [np.min([nt, self._max_output_time_dim])]
```
Comment (matt-long) [inline feisty/offline_driver.py:176]: `days_per_year` ---> `time_levs_per_ds`

File: feisty/offline_driver.py [hunk 10]
```diff
@@ -150,72 +154,101 @@ def __init__(
...
+    # def ds(self):
+    #     """Data comprising the output from ``feisty``."""
+    #     return self._ds
 
-    def _compute_tendency(self, t, state_t):
+    def _compute_tendency(self, forcing_t, state_t):
         """Return the feisty time tendency."""
-        gcm_data_t = self._forcing_t(t)
+        gcm_data_t = self.forcing.interp(time=forcing_t, assume_sorted=True)
+        if not self.allow_negative_forcing:
```
Comment (matt-long) [inline feisty/offline_driver.py:240]: might this logic belong in `feisty_instance`?

File: feisty/offline_driver.py [hunk 11]
```diff
@@ -265,16 +300,59 @@ def run(self, nt, file_out=None, method='euler'):
...
         if file_out is not None:
-            self._ds.to_netcdf(file_out)
+            print(f'Finished _solve and calling _shutdown() at {time.strftime("%H:%M:%S")}')
+            self.gen_ds()
+            self.ds.to_netcdf(file_out)
+            print(f'Finished _shutdown() at {time.strftime("%H:%M:%S")}')
+        else:
+            print(f'Finished _solve at {time.strftime("%H:%M:%S")}')
+
+    def create_ic_file_from_final_state(self, ic_file, overwrite=False):
```
Comment (matt-long) [inline feisty/offline_driver.py:332]:
```suggestion
    def create_restart_file(self, ic_file, overwrite=False):
```

**Review by matt-long — COMMENTED** [review]
Looks great!

**Comment by mnlevy1981** [comment]
[071770d](https://github.com/marbl-ecosys/feisty/pull/30/commits/071770df68c379d976437e063d5d1b69a319806a) cleans up  the straight-forward comments @matt-long made, but the following are NOT yet addressed:

1. better logic surrounding what diagnostics are included in the output
2. moving forcing logic into `feisty_instance`

**Comment by mnlevy1981** [comment]
Also still working on parallelizing `matlab-comparison.ipynb`

**Comment by mnlevy1981** [comment]
[ea12b57](https://github.com/marbl-ecosys/feisty/pull/30/commits/ea12b578be5671d4352155376cf97b08408392dd) addresses #20 because we can't parallelize (part of #21) with global variables in some of the modules

**Comment by mnlevy1981** [comment]
@matt-long I think this is ready for review. The `examples/` directories shows how to set up YAML files to point FEISTY to forcing data, and I'm happy to walk through that with you.

Looking back at previous comments, I'm not sure I ever addressed this:

> [071770d](https://github.com/marbl-ecosys/feisty/pull/30/commits/071770df68c379d976437e063d5d1b69a319806a) cleans up the straight-forward comments @matt-long made, but the following are NOT yet addressed:
> 
>    * better logic surrounding what diagnostics are included in the output
>    * moving forcing logic into feisty_instance

**Pull request merged**